src/main/scala/is/hail/variant/VariantSampleMatrix.scala

```diff
@@ -1246,50 +1250,130 @@ class VariantSampleMatrix[T](val hc: HailContext, val metadata: VariantMetadata,
+          }.map { case (_, ((v, vags), _)) =>
+            (v, vags)
+          }, preservesPartitioning = true)
+
+      case Array(TInterval) =>
+        val intRDD = kt.keyedRDD()
+          .map { case (k, _) => k.getAs[Interval[Locus]](0) }
+          .filter(_ != null)
+          .sortBy(_.start)
+
+        val part = rdd.orderedPartitioner
+        val mappings = intRDD.mapPartitionsWithIndex { case (i, it) =>
+          val overlaps = mutable.Set.empty[Int]
+          it.foreach { interval =>
+            overlaps ++= (part.getPartitionT(interval.start) to part.getPartitionT(interval.end))
+          }
+          overlaps.map(p => (p, i)).iterator
+        }.collect()
+          .groupBy(_._1)
+          .mapValues(_.map(_._2))
 
-    val keyIndex = kt.signature.fieldIdx(kt.key.head)
+        val mArray = mappings.toArray.sortBy(_._1)
 
-    val ktVariantRDD = kt.rdd.map { r =>
-      (r.get(keyIndex).asInstanceOf[Variant], ())
-    }.filter(_._1 != null)
-      .toOrderedRDD
+        if (keep) {
+          if (mArray.length < rdd.partitions.length)
+            info(s"filtered to ${ mArray.length } of ${ rdd.partitions.length } partitions")
+          val zipRDD = new AdjustedPartitionsRDD[Interval[Locus]](intRDD,
+            mArray.map { case (_, intParts) => intParts.map(i => Adjustment[Interval[Locus]](i, identity)) })
 
-    copy(rdd = rdd.orderedLeftJoinDistinct(ktVariantRDD)
-      .mapPartitions(_.filter { case (_, (_, o)) =>
-        Filter.keepThis(o.isDefined, keep)
-      }.map { case (v, (vags, _)) =>
-        (v, vags)
-      }, preservesPartitioning = true)
-      .asOrderedRDD)
+          val subsetRDD = new AdjustedPartitionsRDD[RowT](rdd,
+            mArray.map { case (oldPart, _) => Array(Adjustment[RowT](oldPart, identity)) })
+
+          val newPartitioner = OrderedPartitioner(mArray.init.map { case (oldPart, _) =>
+            rdd.orderedPartitioner.rangeBounds(oldPart)
+          }, mArray.length)
+
+          val filteredRDD = subsetRDD.zipPartitions(zipRDD) { case (it, intervals) =>
+            val itree = IntervalTree.apply[Locus](intervals.toArray)
+            it.filter { case (v, _) => itree.contains(v.locus) }
+          }
+
+          OrderedRDD(filteredRDD, newPartitioner)
+        } else {
+          val zipRDD = new AdjustedPartitionsRDD[Interval[Locus]](intRDD,
+            (0 until nPartitions).map { i =>
+              val intParts = mappings.getOrElse(i, Array.empty[Int])
+              intParts.map(i => Adjustment[Interval[Locus]](i, identity))
+            })
+
+          rdd.zipPartitions(zipRDD, preservesPartitioning = true) { case (it, intervals) =>
+            val itree = IntervalTree.apply[Locus](intervals.toArray)
+            it.filter { case (v, _) => !itree.contains(v.locus) }
+          }
+        }
+
+      case _ => fatal(
+        s"""method 'filterVariantsTable' requires a table with one of the following keys:
+           |  [ Variant ]
```

Comment:
The key can include multiple columns, the bracket indicates that there's only one column and it's a variant. Can change this if you prefer no brackets, but since we print the list of key columns with brackets, I think this is consistent.